File: examples/02-reliability/go/heartbeat_resume_from_checkpoint.go
package reliability

import (
	"context"

	"go.temporal.io/sdk/activity"
)

// Real-world: a multi-hour data backfill that must resume where it left off
// after a worker crash or activity retry — not restart from page 0. The
// heartbeat carries the last completed page; on a retry the next attempt reads
// it back (activity.GetHeartbeatDetails) and skips the work already done.
func Backfill(ctx context.Context, dataset string) (string, error) {
	startPage := 0
	if activity.HasHeartbeatDetails(ctx) {
		_ = activity.GetHeartbeatDetails(ctx, &startPage) // resume point
	}

	for page := startPage; page < 100_000; page++ {
		copyPage(dataset, page)
		activity.RecordHeartbeat(ctx, page) // checkpoint: this page is done
	}
	return "backfilled " + dataset, nil
}

File: examples/02-reliability/java/heartbeat_resume_from_checkpoint.java
// Real-world: a multi-hour data backfill that must resume where it left off
// after a worker crash or activity retry — not restart from page 0. The
// heartbeat carries the last completed page as its "details"; on a retry the
// next attempt reads that detail back and skips the work already done.
class BackfillActivitiesImpl implements BackfillActivities {
  @Override
  public String backfill(String dataset) {
    ActivityExecutionContext ctx = Activity.getExecutionContext();

    // Resume point: the details attached to the most recent heartbeat of the
    // PREVIOUS attempt, if this is a retry. Empty on the first attempt.
    int startPage =
        ctx.getHeartbeatDetails(Integer.class).orElse(0);

    for (int page = startPage; page < 100_000; page++) {
      copyPage(dataset, page);
      ctx.heartbeat(page); // checkpoint: this page is done
    }
    return "backfilled " + dataset;
  }
}
